### DISCUSSION TITLE
Should return type inference infer literals ever?

### DISCUSSION THREAD

**Discussion by jbrunton96** [microsoft/pyright#7158] [General]
When trying to run Pyright on an existing codebase, I noticed it complained about [the following code](https://pyright-play.net/?code=MYGwhgzhAECCBcAoaLoBMCmAzaWD2eAFBBiFgJRKrXQBOGALgK60B20AYmCCcqnykw4ARmFrFSFKjRT1mbaACIAEqRB5oAdzy0QaRYkOhIMAEKFYlAemy4CEslZmzGLdgBVaTDIepDoouIkjtIycm5KAOIEaMIAnhgGQA):

```python
class A:
    def foo(self):
        return False
    
    def bar(self):
        return "Hello world"


class B(A):
    def foo(self):  # Return type mismatch: base method returns type "Literal[False]", override returns type "Literal[True]"
        return True

    def bar(self):  # Return type mismatch: base method returns type "Literal['Hello world']", override returns type "Literal['Goodbye']"
        return "Goodbye"
```

It's a relatively common pattern for subclasses to override a method to return a different value, but Pyright infers the return types of the base implementations as `Literal[False]` and `Literal["Hello world"]` here. I can see exactly why it's doing that - it is the most specific return type that it can give, but I'm not sure it's very useful behaviour. 

Obviously I can stop the error here by annotating the return type as `-> bool` and `-> str`, but is it ever useful a return type to be inferred to be a `Literal` instead of a type? I can't think of any use-cases off the top of my head where you'd want to know that a function returns a specific literal, other than the special case of `None` I guess. 

Is there a use-case for this behaviour that I'm not thinking of? Should Pyright explicitly walk up to the type when inferring return types?

**Comment by erictraut** [comment]
This is working as designed. There are both pros and cons of inferring literal types. The pros generally outweigh the cons, especially when you consider language server features. For that reason, we're unlikely to change the current behavior.

If you want to override the standard inference behaviors, you should provide explicit return type annotations.
